### PR TITLE
feat: doesObjectExist and doesBucketExist v2

### LINKED ISSUE #2342: doesObjectExist should not return true or false based on assumptions
Confirm by changing [ ] to [x] below to ensure that it's a bug:
- [x] I've gone though [Developer Guide](https://docs.aws.amazon.com/sdk-for-php/v3/developer-guide/welcome.html) and [API reference](https://docs.aws.amazon.com/aws-sdk-php/v3/api/index.html)
- [x] I've checked [AWS Forums](https://forums.aws.amazon.com) and [StackOverflow](https://stackoverflow.com/questions/tagged/aws-php-sdk) for answers
- [x] I've searched for [previous similar issues](https://github.com/aws/aws-sdk-php/issues) and didn't find any solution
  
**Describe the bug**
If the request made in `doesObjectExist` fails for any reason, it should throw an exception instead of returning `false` based on any assumptions (e.g. bucket does not exist if endpoint can not be reached). Otherwise, the result may be wrong, i.e. `false` is returned, even though the object does actually exist.

I guess it is a [similar issue](https://github.com/aws/aws-sdk-java/issues/1256) as in the Java SDK. You may consider to introduce `doesBucketExistV2` [as well](https://github.com/aws/aws-sdk-java/issues/1256#issuecomment-331214701)

> Due to some concerns about internal service teams being tightly coupled to the underlying API call we decided to rollback this change so that doesBucketExist continues to use headBucket. We've deprecated this method and introduced a doesBucketExistV2 which uses the new getBucketAcl approach.

**Version of AWS SDK for PHP?**
Example: v3.191.6

**Version of PHP (`php -v`)?**
7.4

**To Reproduce (observed behavior)**
To use an endpoint URL that can not be resolved, is IMO the easiest way to get an error in the request. Otherwise, it depends on your endpoint configuration. You can run the following example without any changes to the configuration:

```php
$client = new \Aws\S3\S3Client([
  'version' => 'latest',
  'region' => 'eu-central-1',
  'endpoint' => 'http://endpoint-that-can-not-be-resolved',
  'use_path_style_endpoint' => true,
  'credentials' => [
    'key' => 's3CredentialsKey',
    'secret' => 's3CredentialsSecret',
  ],
]);

$bucketName = 'bucketName';
$doesExist = $client->doesObjectExist($bucketName, 'bucketKey');
// should not be reached, since exception should be thrown by doesObjectExist:
var_dump($doesExist); // false
// the following throws an exception as expected
$client->createBucket(['Bucket' => $bucketName]);
```

You can also reproduce this issue with your own endpoint if you use invalid credentials to trigger a 403. In that case, `doesObjectExist` should throw an exception, too.

**Expected behavior**
`doesObjectExist` should throw an exception.

**Additional context**
In my private project, I had an issue, when I uploaded a lot of files. I used `doesObjectExist` to check if I already uploaded the file. This case was actually meant to be used, if I have to run the script again. However, in the first run, `doesObjectExist` returned `true` in some cases (0.1%), even though the file did actually not exist and hence, wasn't uploaded. Are there any assumptions, where `true` is returned? I'm not sure how to reproduce this consistently or how to log relevant information if the result was wrong, but I guess that requests have failed with another error as I described in the reproduce section of this issue.

### PR DESCRIPTION
**Pull request by stobrien89**
*Issue #, if available:*
Implements #1561, closes #2342 

*Description of changes:*

New object/bucket existence helper methods with changed exception handling behavior.  Previously, `false` was returned any time an exception was thrown by the underlying S3 APIs (`headObject` and `headBucket`) unless the exception's status code was >= `500` (then exception was thrown) or the exception's error code was `AccessDenied` (`true` was returned).  This was problematic because `false` could be returned in the cases of an unresolvable endpoint, expired credentials, invalid credentials, or if bucket permissions were not granted to the calling entity and resources did in fact exist.

This implementation makes fewer assumptions in cases where status codes are ambiguous, namely `403` status codes. a `403` status code can signify bad credentials or incorrect bucket-level permissions— this causes issues with `headObject`/`doesObjectExist` calls in particular because a 403 will be returned whether or not an object exists. In this implementation, exceptions will be thrown in all cases when an exception is thrown by the underlying API, unless the exception has a  `404` status code (returns false) or in the case of `DoesBucketExist`, a user can specify if they want to accept `403` status codes, as a 403 will be thrown if a bucket exists, but an entity doesn't have correct bucket-level permissions.

Other changes include accepting bucket redirects (incorrectly specified region) as a `true` case for `DoesBucketExist`, the `$accept403` flag for `doesBucketExist` and the `$includeDeleteMarkers` flag for `doesObjectExist`, which counts a `404` response with the presence of a delete marker header as a `true` case.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/S3/S3ClientTrait.php [hunk 1]
```diff
@@ -297,6 +330,42 @@ private function checkExistenceWithCommand(CommandInterface $command)
...
+        CommandInterface $command,
+        $accept403,
+        $includeDeleteMarkers = false
+    )
+    {
+        try {
+            $this->execute($command);
+            return true;
+        } catch (S3Exception $e) {
+            if (($accept403 && $e->getStatusCode() === 403)
```
[the remark this answers is not in the document]
Comment (SamRemis) [reply src/S3/S3ClientTrait.php:353]: Another option (that I like less well cause of performance) would be to add some variables above this like so:

```
$statusCode = $e->getStatusCode();
$exceptionHasDeleteMarker = !empty($e->getResponse()->getHeader("x-amz-delete-marker"))
if (
       ($accept403 && $statusCode == 403)
       || ($includeDeleteMarkers && $exceptionHasDeleteMarker)
 ) {
       return true;
 }
```

File: src/S3/S3ClientInterface.php [hunk 2]
```diff
@@ -48,6 +48,17 @@ public function getObjectUrl($bucket, $key);
      */
     public function doesBucketExist($bucket);
 
+    /**
+     * Determines whether or not a bucket exists by name.
+     *
+     * @param string $bucket  The name of the bucket
+     * @param bool $accept403 Changes behavior to return true in the case of a 403.
```
Comment (SamRemis) [inline src/S3/S3ClientInterface.php:55]: How about "set to true to have this method return true in the case of a 403 credential error; using the default value of false will cause the error to be thrown instead"

File: src/S3/S3ClientTrait.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (stobrien89) [reply src/S3/S3ClientTrait.php:353]: I wasn't sure how we felt about separate if statements that result in the same return value, but I like that approach.  Better for readability and less cumbersome.

File: src/S3/S3ClientTrait.php [hunk 3]
```diff
@@ -297,6 +330,42 @@ private function checkExistenceWithCommand(CommandInterface $command)
...
+                || (!empty($e->getResponse()->getHeaders()["x-amz-delete-marker"])
```
[the remark this answers is not in the document]
Comment (stobrien89) [reply src/S3/S3ClientTrait.php:354]: Agreed. I'll get that updated

File: src/S3/S3ClientTrait.php [hunk 4]
```diff
@@ -297,6 +330,42 @@ private function checkExistenceWithCommand(CommandInterface $command)
...
+                    && $includeDeleteMarkers)
+                || ($command->getName() === 'HeadBucket'
```
[the remark this answers is not in the document]
Comment (stobrien89) [reply src/S3/S3ClientTrait.php:356]: This is a condition I'll need to test again— the error code _should_ be a 301, but I'm pretty sure it wasn't exposed on the exception itself and it was shorter to write an `instanceof` than opening up the response for it.

File: src/S3/S3ClientTrait.php [hunk 5]
```diff
@@ -260,6 +261,30 @@ public function doesBucketExist($bucket)
...
+     */
+    public function doesBucketExistV2($bucket, $accept403 = false)
+    {
+        $command = $this->getCommand('HeadBucket', ['Bucket' => $bucket]);
+
+        try {
+            $this->execute($command);
+            return true;
+        } catch (S3Exception $e) {
+            if (($accept403 && $e->getStatusCode() === 403)
```
Comment (SamRemis) [inline src/S3/S3ClientTrait.php:275]: there should be a newline after the first parenthesis

File: src/S3/S3ClientTrait.php [hunk 6]
```diff
@@ -260,6 +261,30 @@ public function doesBucketExist($bucket)
...
+     * @see S3ClientInterface::doesBucketExistV2()
+     */
+    public function doesBucketExistV2($bucket, $accept403 = false)
+    {
+        $command = $this->getCommand('HeadBucket', ['Bucket' => $bucket]);
+
+        try {
+            $this->execute($command);
+            return true;
+        } catch (S3Exception $e) {
```
Comment (SamRemis) [inline src/S3/S3ClientTrait.php:274]: if it's a different kind of exception, it'll just get thrown back to the user, right?

File: src/S3/S3ClientTrait.php [hunk 7]
```diff
@@ -273,6 +298,42 @@ public function doesObjectExist($bucket, $key, array $options = [])
...
+            $response = $e->getResponse();
+
+            if ($includeDeleteMarkers
+                && (!empty($response)
+                    && $response->getHeader('x-amz-delete-marker'))
+            )
+            {
```
Comment (SamRemis) [inline src/S3/S3ClientTrait.php:327]: this should be on the line above: `) {`

File: src/S3/S3ClientTrait.php [hunk 8]
```diff
@@ -273,6 +298,42 @@ public function doesObjectExist($bucket, $key, array $options = [])
...
+        );
+
+        try {
+            $this->execute($command);
+            return true;
+        } catch (S3Exception $e) {
+            $response = $e->getResponse();
+
+            if ($includeDeleteMarkers
+                && (!empty($response)
```
Comment (SamRemis) [inline src/S3/S3ClientTrait.php:324]: Maybe extract this into a hasDeleteMarker method that checks if a response exists and returns whether or not that header is there?  Condense a a bit of code.

File: src/S3/S3ClientTrait.php [hunk 9]
```diff
@@ -273,6 +298,42 @@ public function doesObjectExist($bucket, $key, array $options = [])
...
+    /**
+     * @see S3ClientInterface::doesObjectExistV2()
+     */
+    public function doesObjectExistV2(
+        $bucket,
+        $key,
+        $includeDeleteMarkers = false,
+        array $options = []
+    )
+    {
```
Comment (SamRemis) [inline src/S3/S3ClientTrait.php:310]: should be on the same line as the parenthesis

File: src/S3/S3ClientTrait.php (hunk 6 above)
Comment (stobrien89) [reply src/S3/S3ClientTrait.php:274]: yeah, we're only interested in inspecting S3exceptions for true/false conditions and the rest can be thrown back to the user.

File: src/S3/S3ClientTrait.php (hunk 8 above)
Comment (stobrien89) [reply src/S3/S3ClientTrait.php:324]: totally agree. just took a second glance and that looks pretty ugly

**Review by SamRemis — APPROVED** [review]

**Comment by SimonMacIntyre** [comment]
I'm confused by this changelog, it says `doesBucketExistV2` was deprecated and to use `doesObjectExistV2` instead, is that a mistake?

`doesBucketExistV2` is a brand new method in this PR, so I assume it was not intended to implement a brand new method and deprecate it in the same merge?

image

**Pull request merged**